### PR TITLE
UI\Component\Dropdown no item rendering

### PR DESCRIPTION
**Pull request by dkloepfer**
Checking items passed at construction and render itemless dropdowns as inactive

### CODE REVIEW

File: src/UI/Implementation/Component/Dropdown/Dropdown.php [hunk 1]
```diff
@@ -87,4 +88,22 @@ public function withOnHover(Signal $signal) {
...
+	 */
+	protected function checkItems(array $items)
+	{
+		foreach ($items as $item) {
+			if(!(	$item instanceof \ILIAS\UI\Component\Button\Shy
+				||	$item instanceof \ILIAS\UI\Component\Divider\Horizontal)) {
+					throw new \InvalidArgumentException('invalid item');
+			}
+		}
+	}
```
Comment (klees) [inline src/UI/Implementation/Component/Dropdown/Dropdown.php]: This is already implemented as `ILIAS\UI\Implementation\Component\ComponentHelper::checkArgListElements`. Please use that implementation.

**Comment by klees** [comment]
This is something I stumbled upon when implementing #1054. The component contain some imo undocumented and rather implicit behaviour, which was "if no actions are passed, don't render". This is the attempt to turn that implicit behaviour into something more apparent by rendering a dropdown without actions as disabled.

@Amstutz What do you think of this in general?

**Review by klees — CHANGES_REQUESTED** [review]
Please also add some according lines to the KS-section of the dropdown. @nhaagen will help you, if required.

**Review by Amstutz — CHANGES_REQUESTED** [review]
I am not sure if this is any good. If not actions are there why rendering the dropdown at all? I believe we should keep the UI as clean as possible and not cluster it with items, without any use. We could also state this in an rule but in which case then would you render an empty dropdown? What would be the use of it?

**Comment by dkloepfer** [comment]
@Amstutz The one usecase we had in mind is an array of autogenerated dropdowns (e.g. in a table). It should be apparent to a user that there usually is a selection. Also, if there is no selection for one of the dropdowns, an ugly gap in the list is formed if simply nothing is rendered.

**Comment by Amstutz** [comment]
@dkloepfer Thanks for the feedback. I am still not convinced, that this is right. Probably we have to forces at work here. Consistency, which is valued high in this case by @dkloepfer and affordance which is my argument. Both are very basic and commonly accepted UI principles (see https://www.csun.edu/science/courses/671/bibliography/preece.html). 

However, my opinion is of minor importance in this mather. I will ask Yvonne about it. Maybe you could also ping christoph? If they both believe it to be necessary, I'll be convinced.

**Comment by dkloepfer** [comment]
@catkrahl Any opinions on this matter?

**Comment by dkloepfer** [comment]
@Amstutz Of course consitency and affordance could be satisfied simultaniously by adding a new dropdown-element, let's call it persistentDropdown momentarily, just for the sake of argument, which is allways rendered (allthough inactive at times). However, i don't know, wether this will nicely fit the UI framework paradigms. Any ideas on this.

**Comment by Amstutz** [comment]
I object rendering inactive drop downs anywhere, also in tables. Maybe there are other cases where they might be useful, however I do not see them yet. So atm, this would not solve my issue. Note that this might be just my issue. Let's wait for some more feedback.

**Comment by yvseiler** [comment]
As always, it depends on the context in which the element is used. From my perspective, I would use the following differentiator to show the dropdown (or other buttons) in a disabled state or hide it:

1. **Disable**: The item is activated when the user performs a specific action. (e.g. Submit button: A mandatory field is not yet filled in in the form, if I fill in the field, the Submit button is displayed)
2. **Hide**: The element cannot be changed to an active status by a user action. (e.g. a button is not visible because the user does not have sufficient rights to do something with it).

For the dropdown in the table described here, I therefore believe that it depends on this:
1. As a user I can change something on the screen of the table and the dropdown becomes usable **-> then deactivate**.
2. As a user I cannot influence the availability of the drop-down on the table screen _-> then hide_.

In addition to 1), it should be explained why the element cannot currently be used (communicate why it is disabled).

Currently I understand the dropdown in this request to mean that I can't select anything in the Actions menus because I don't have enough permissions. Therefore I tend to "hide".

_Of course, if this dropdown is not rendered, the table must not be distorted visually. Above there is mentioned a "gap". The layout should be able to adapt accordingly, depending on whether dropdowns are displayed or not. Unfortunately, I have no visual idea of such a "gap", so I can only imagine the effects._

**Comment by klees** [comment]
@Amstutz I think you are slightly missing the issue I intended to be solved.

Currently, the Renderer of the dropdown contains [these lines](https://github.com/ILIAS-eLearning/ILIAS/blob/release_5-3/src/UI/Implementation/Component/Dropdown/Renderer.php#L30). The implemented behaviour ("do not render dropdowns without actions") is currently undocumented and arguably shouldn't be located in the renderer. To me this violates the principle of least surprise, which I consider extremly useful when designing APIs (and other stuff?)

I see three ways to solve this:
1. Document the behaviour. This would require some changes in the rules.
2. Remove the behaviour by forbidding empty action lists (via Exception).
3. Remove the behaviour by rendering a none functional dropdown.

3. is what we proposed here. 2. would be fine for me as well. 1. to me would be slightly odd because we currently have no precedence for similar behaviour ("do not render some component for some reason"). 1. would mean that I could have situations, where I clearly see that a dropdown is created but I never see it on the GUI, especially when also the list of actions is created somewhere else. I would rather like to see an according `if` in that code than in the GUI.

The stuff Yvonne is writing is fine for me as well but I cannot see how the UI-framework could possibly decide on that issue. Which is close to the reason I do not like option 1.

**Comment by Amstutz** [comment]
@klees No I am not missing this, I just disagree with the proposed solution. Howver, thanks for your effort to solve this.

1. Is fine for me (however not the option I prefer, see bellow). This is exactly why we do have those written descriptions and rules. I would just add a note in the composition section, not the rules. Something like: "Note that empty dropdowns are not rendered at all to keep the UI as clean as possible". 

2. Would be fine for me as well. Probably even better, since we prevent some unneeded logic got be executed and make the intend even clearer (as you point out). However, 2 would need a explicit rule IMO to make the reasoning transparent. Something like above would do: "Dropdowns MUST cointain items. Empty dropdowns are not allowed to keep the UI as clean as possible."

3. I do not like.

4. Is what I prefer with some additonal sugar. Clearly state that if an empty dropdown is created, it will be disabled. Such dropdowns MUST only be used, if the user has some possibility to change something on the screen to make some action in the dropdown available." (Yvonnes Feedback).

> would mean that I could have situations, where I clearly see that a dropdown is created but I never see it on the GUI, especially when also the list of actions is created somewhere else. I would rather like to see an according if in that code than in the GUI.

Resonates with me. Thanks for the effort for fixing this.

> The stuff Yvonne is writing is fine for me as well but I cannot see how the UI-framework could possibly decide on that issue. Which is close to the reason I do not like option 1.

You are right, the UI framework can not decide on this. However this is crucially important knowhow, on which the developer has to be made aware. This is exactly why we have the written rule. It will not possible to check on everything by logic (however you are right in being cumbersome in always pointing out that we should as much as possible). This is why I prefer some sugared version of your proposal.

**Comment by klees** [comment]
@dkloepfer 
Please promote the current change in the rules of the `Standard Dropdown` on node higher in the hierarchy to `Dropdown`. Please add a usage rule that states that "Dropdowns containing no interactions will be rendered as disabled. Such dropdowns MUST only be used, if the user has some possibility to change something on the screen to make some action in the dropdown available."

Thx!

**Review by klees — APPROVED** [review]

**Review by Amstutz — APPROVED** [review]
Thanks a lot.

**Review by alex40724 — CHANGES_REQUESTED** [review]
This changes the default behaviour now for all existing DropDowns, right? So now all the disabled DropDowns will appear. Like Timon I think that showing a disabled DropDown will be the exception, but now it becomes the default behaviour. This means all consumer code that relied on the old behaviour now needs to add these "if (count($items))". So we are moving a lot of "ifs" into consumer code?

**Comment by klees** [comment]
Yes @alex40724, this is correct. Currently there are 5 usage sides outside the UI-framework for the dropdown, 4 inside the UI-framework and some more in the examples. I do not know if any of these would require an if, though.

**Comment by matthiaskunkel** [comment]
Jour Fixe, 18 JUN 2018 : We prefer solution 1 - just adding a note to the guideline that dropdowns without items are not rendered. This avoids certain refactoring in consumer classes. Timon will offer an alternative suggestion.

**Pull request closed without merging**